main/test/ca/uwaterloo/flix/library/TestRedBlackTree.flix

```diff
@@ -190,32 +190,69 @@ namespace TestRedBlackTree {
 
 
     /////////////////////////////////////////////////////////////////////////////
-    // map                                                                     //
+    // mapWithKey                                                              //
     /////////////////////////////////////////////////////////////////////////////
 
     @test
-    def map01(): Bool =
+    def mapWithKey01(): Bool =
         let Map(t) = Map#{}: Map[Unit, Int32];
-        Map(RedBlackTree.mapWithKey((_, v) -> v, t)) == Map#{}
+        Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == Map#{}
 
     @test
-    def map02(): Bool =
+    def mapWithKey02(): Bool =
         let Map(t) = Map#{1 -> 4};
         Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == Map#{1 -> 5}
 
     @test
-    def map03(): Bool =
+    def mapWithKey03(): Bool =
         let Map(t) = Map#{2 -> -1, 0 -> 0};
         Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 0 -> 0}
 
     @test
-    def map04(): Bool =
+    def mapWithKey04(): Bool =
         let Map(t) = Map#{2 -> -1, 5 -> 15, 11 -> -9};
         Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 5 -> 20, 11 -> 2}
 
     @test
-    def map05(): Bool =
+    def mapWithKey05(): Bool =
         let Map(t) = Map#{2 -> -1, 5 -> 15, 11 -> -9, 8 -> 8};
         Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 5 -> 20, 11 -> 2, 8 -> 16}
 
-}
\ No newline at end of file
+    @test
+    def mapWithKey06(): Bool =
+        let Map(t) = List.zip(List.range(0, 2000), List.range(0, 2000)) |> List.toMap;
+        let expected = List.zip(List.range(0, 2000), List.range(1, 2001)) |> List.toMap;
+        Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == expected
+
+    @test
+    def mapWithKey07(): Bool =
+        let Map(t) = List.zip(List.range(0, 2000), List.range(-1000, 1000)) |> List.toMap;
+        let expected = List.zip(List.range(0, 2000), List.range(-999, 1001)) |> List.toMap;
+        Map(RedBlackTree.mapWithKey((k, v) -> k + v, t)) == expected
+
+    @test
+    def mapWithKey08(): Bool & Impure =
+        let range = List.range(0, 2000);
+        let Map(t) = List.zip(range, range) |> List.toMap;
+        let a = [2000; default];
+        RedBlackTree.mapWithKey((k, v) -> { a[k] = v; v }, t);
```

Comment:
If this was done in parallel (which would be wrong because if impurity) your test would still pass. The problem is that updating the array with a[k] cannot go wrong, right? Even if done in parallel. I think what you should do is store each key into a mutlist/array and check that it occurs in the right order.

Reply:
We really want an operation that is not associative nor commutative.

For example, if write or append elements to a mutable array then doing it in parallel would completely destroy the order of the array which is the behavior we want to trigger if there is a bug.

Reply:
Does it make sense? Perhaps we can take a quick zoom?

Reply:
Alright, I've pushed some changes now. If they're not testing for the right thing, then I think I need some more help with this. I've tried to do operations that would be prone to race conditions as you suggested :safety_vest: